xla/hlo/analysis/hlo_replication_analysis_test.cc

```diff
@@ -789,6 +795,177 @@ ENTRY entry {
       replica_groups0));
 }
 
+TEST_F(
+    HloReplicationAnalysisTest,
+    PartiallyReplicatedAllGatherFlattenedIDPartitionAnalysisAsymmetricGroups) {
+  const std::string module_str = R"(
+HloModule GlobalIdAllGather
+
+ENTRY entry {
+  param = f32[1] parameter(0)
+  ROOT all_gather = f32[6] all-gather(param), replica_groups={{0,1,2,3,6,7},{4,5,8,9,10,11}}, dimensions={0}, use_global_device_ids=true, channel_id=1
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(
+      auto module, ParseAndReturnVerifiedModule(module_str, /*replica_count=*/2,
+                                                /*num_partitions=*/6));
+  TF_ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<HloReplicationAnalysis> partition_analysis,
+      HloReplicationAnalysis::RunWithPartialReplication(
+          module.get(),
+          /*cross_partition_spmd=*/true));
+
+  std::array<ReplicaGroup, 3> replica_groups0;
+  replica_groups0[0].add_replica_ids(0);
+  replica_groups0[0].add_replica_ids(1);
+  replica_groups0[1].add_replica_ids(2);
+  replica_groups0[1].add_replica_ids(3);
+  replica_groups0[2].add_replica_ids(4);
+  replica_groups0[2].add_replica_ids(5);
+
+  std::array<ReplicaGroup, 2> replica_groups1;
+  replica_groups1[0].add_replica_ids(0);
+  replica_groups1[0].add_replica_ids(1);
+  replica_groups1[0].add_replica_ids(2);
+  replica_groups1[1].add_replica_ids(3);
+  replica_groups1[1].add_replica_ids(4);
+  replica_groups1[1].add_replica_ids(5);
+
+  EXPECT_TRUE(partition_analysis->HloInstructionIsReplicatedAt(
+      FindInstruction(module.get(), "all_gather"), {}, replica_groups0));
+  EXPECT_FALSE(partition_analysis->HloInstructionIsReplicatedAt(
+      FindInstruction(module.get(), "all_gather"), {}, replica_groups1));
+}
+
+TEST_F(
+    HloReplicationAnalysisTest,
+    PartiallyReplicatedAllGatherFlattenedIDReplicationAnalysisAsymmetricGroups) {
+  const std::string module_str = R"(
+HloModule GlobalIdAllGather
+
+ENTRY entry {
+  param = f32[1] parameter(0)
+  ROOT all_gather = f32[6] all-gather(param), replica_groups={{0,1,2,3,4,6},{5,7,8,9,10,11}}, dimensions={0}, use_global_device_ids=true, channel_id=1
```

Comment:
The added functions are all almost identical. To avoid verbosity, I would have have a helper function `CollectiveIsReplicated` so you could do, e.g.

```c++
EXPECT_FALSE(CollectiveIsReplicated(
    /*collective_replica_groups=*/{{0,1,2,3,4,6},{5,7,8,9,10,11}}, 
    /*replica_count=*/6, 
    /*num_partitions=*/2, 
    /*cross_partition_spmd=*/false, 
    /*analysis_replica_groups=*/{{0,1},{2,3},{4,5}}));
```